tests/parser/functions/test_convert.py

```diff
@@ -1,544 +1,532 @@
+    elif typename.startswith("String"):
+        assert typename == "String[32]"  # TODO test others
+        return TestType(typename, 32, "Bytes", info)
+    elif typename == "address":
+        return TestType(typename, 20, "address", None)
+    elif typename == "bool":
+        return TestType(typename, 1, "bool", None)
 
-    return type_
+    raise AssertionError(f"no info {typename}")
 
 
-def can_convert(o_typ, i_typ):
+def can_convert(i_typ, o_typ):
     """
     Checks whether conversion from one type to another is valid.
     """
-    it = _get_case_type(i_typ)
-    ot = _get_case_type(o_typ)
-
-    # Skip bytes20 because they are treated as addresses
-    if i_typ == "bytes20":
+    if i_typ == o_typ:
         return False
 
-    # Check
-    if ot == "bool":
-        return it in ["int", "uint", "decimal", "bytes", "Bytes", "address"]
+    i_detail = _parse_type(i_typ)
+    o_detail = _parse_type(o_typ)
 
-    elif ot == "int":
-        return it in ["uint", "decimal", "bytes", "Bytes", "bool"]
+    if o_typ == "bool":
+        return True
+    if i_typ == "bool":
+        return o_typ not in {"address"}
 
-    elif ot == "uint":
-        return it in ["int", "decimal", "bytes", "Bytes", "address", "bool"]
+    if i_detail.type_class == "int":
+        if o_detail.type_class == "bytes":
+            return i_detail.type_bytes <= o_detail.type_bytes
 
-    elif ot == "decimal":
-        return it in ["int", "uint", "bytes", "Bytes", "bool"]
+        ret = o_detail.type_class in ("int", "decimal", "bytes", "Bytes")
+        if not i_detail.info.is_signed:
+            ret |= o_typ == "address"
+        return ret
 
-    elif ot == "bytes":
-        ot_bytes_info = parse_bytes_m_info(o_typ)
-        if it in ["int", "uint"]:
+    elif i_detail.type_class == "bytes":
+        if o_detail.type_class in ("int", "Bytes", "address"):
             # bytesN must be of equal or larger size to [u]intM
-            it_int_info = parse_integer_typeinfo(i_typ)
-            return ot_bytes_info.m_bits >= it_int_info.bits
-        elif it == "Bytes":
-            # bytesN must be of equal or larger size to Bytes[M]
-            return ot_bytes_info.m >= _get_type_N(i_typ)
-        elif it == "address":
-            return ot_bytes_info.m_bits >= ADDRESS_BITS
-        return it in ["decimal", "Bytes", "address"]
+            return i_detail.type_bytes <= o_detail.type_bytes
+
+        # the fact that any bytes can be converted to decimal but not all ints
+        # may be an inconsistency in the spec
+        return i_detail.type_class in ("decimal", "bytes")
+
+    elif i_detail.type_class == "Bytes":
+        return o_detail.type_class in ("int", "decimal", "address")
+
+    elif i_typ == "decimal":
+        if o_detail.type_class == "bytes":
+            return i_detail.type_bytes <= o_detail.type_bytes
+
+        return o_detail.type_class in ("int", "bool")
+
+    elif i_typ == "address":
+        return o_typ in ("uint", "bytes")
```

Comment:
```suggestion
        if o_detail.type_class == "bytes":
            return i_detail.type_bytes < o_detail.type_bytes
        elif o_detail.type_class == "int":
            return not o_detail.info.is_signed
        return False
```
We need to filter for `bytesN` where `N` > 20, and for `int` in the case of "int" type_class.

Reply:
why is that?